fix DATAJPA-1822

pblanchardie opened the pull request:
<!--

Thank you for proposing a pull request. This template will guide you through the essential steps necessary for a pull request.
Make sure that:

-->

- [X] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [X] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAJPA).
- [X] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [X] You submit test cases (unit or integration tests) that back your changes.
- [X] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

pivotal-issuemaster commented:
@pblanchardie Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-jpa&pullRequestId=436)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-jpa&pullRequestId=436) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@pblanchardie Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-jpa&pullRequestId=436)!

pblanchardie commented:
Hi @schauder, I would appreciate some feedback and don't hesitate to ask for clarification.

schauder reviewed (changes requested):
Two integration tests fail with EclipseLink:

Please make sure to run all tests with `./mvnw clean verify`

```
[ERROR] Failures: 
[ERROR]   EclipseLinkQueryUtilsIntegrationTests>QueryUtilsIntegrationTests.createsLeftJoinForNonOptionalToOneWithNestedOptional:147 
Expecting empty but was: [org.eclipse.persistence.internal.jpa.querydef.JoinImpl@56bb4183]
[ERROR]   EclipseLinkQueryUtilsIntegrationTests>QueryUtilsIntegrationTests.createsLeftJoinForOptionalToOneWithNestedNonOptional:131 
Expecting empty but was: [org.eclipse.persistence.internal.jpa.querydef.JoinImpl@7b1248b8]
```

pblanchardie commented:
> Two integration tests fail with EclipseLink:

Temporarily fixed by skipping these tests for EclipseLink.

This is caused by [QueryUtils.java#L699](https://github.com/spring-projects/spring-data-jpa/blob/4bc11ac76c864e8373f39e3752970e5913146056/src/main/java/org/springframework/data/jpa/repository/query/QueryUtils.java#L699) producing an inner join.
Looks like [EclipseLink Bug 413892](https://bugs.eclipse.org/bugs/show_bug.cgi?id=413892), but also occurs with EclipseLink 2.7.5.

pblanchardie commented:
Added a workaround for EclipseLink (inspired by the existing workaround using `ManagedType`) that also handles intermediate joins.
Updated the comment from 2013 saying that the workaround is needed for EclipseLink 2.4.x to "still occurs as of 2.7".
Should we add a TODO to plan the removal of these ugly workarounds?
Should we reopen a bug in EclipseLink?

schauder commented:
> Should we reopen a bug in EclipseLink?

Thanks for bringing this to the attention of the EclipseLink team. 
But it seems this is more of a Spec question now:
https://github.com/eclipse-ee4j/jpa-api/issues/169

schauder commented:
Thanks, this is merged, polished and back ported.
